Sources/MixpanelInstance.swift

```diff
@@ -434,61 +431,92 @@ open class MixpanelInstance: CustomDebugStringConvertible, FlushDelegate, AEDele
         if hasOptedOutTracking() {
             return
         }
+
+        // Start the background task on DidEnterBackground so that flush/archive have plenty
+        // of time to execute
         
         taskId = sharedApplication.beginBackgroundTask { [weak self] in
-            self?.taskId = UIBackgroundTaskIdentifier.invalid
+
+                // If self DNE, then we're already dead and can't do anything anyways
+                // This shouldn't happen in practice
+                guard let self = self else { return }
+
+                // Expiration handler: All Code inside this block MUST BE SYNCHRONOUS!
+                // This block is the system telling us we're about to be ejected, so we need to
+                // 1) Always end the task
+                // 2) set the taskID to invalid
+                // Any async calls here will never be executed (at best) or crash at worst.
+
+                #if DECIDE
+                self.readWriteLock.write {
+                    self.decideInstance.decideFetched = false
+                }
+                #endif
+
+                // If the system is calling the expiration handler, the taskID must be valid (?)
+                // Either way, there is no known harm in ending an invalid background task
+                // It is also possible iOS has some internal hack requiring a call to endBackgroundTask
+                // always, so by "defensively" calling endBackgroundTask here we are insuring iOS
+                // doesn't kill our process. Note that repeated "violations" of not ending your task
+                // will influence the background time iOS grants you in future requests.
+                sharedApplication.endBackgroundTask(self.taskId)
+                self.taskId = UIBackgroundTaskIdentifier.invalid
```

Comment:
I think you are spot on the issues for the background task. I will make the change based on your suggestion. Thanks!